File: 0739-daily-temperatures/0739-daily-temperatures.py
class Solution:
    def dailyTemperatures(self, temperatures: List[int]) -> List[int]:
        stack = []
        res = [0] * len(temperatures)
        
        # traverse the list
        for i in range(len(temperatures)):
            cur_temp = temperatures[i]
            # Empty stack or not a warmer day
            if len(stack) == 0 or cur_temp <= stack[-1][0]:
                stack.append((cur_temp, i))
            # Found next warmer day
            else:
                while stack and stack[-1][0] < cur_temp:
                    prev_temp, index = stack.pop()
                    res[index] = i - index
                stack.append((cur_temp, i))
        
        return res

File: 0739-daily-temperatures/0739-daily-temperatures.cpp
class Solution {
public:
    vector<int> dailyTemperatures(vector<int>& temperatures) {
        vector<int> res(temperatures.size(), 0);
        stack<pair<int, int>> stk;

        int index;
        for (int i = 0; i < temperatures.size(); i++) {
            while (!stk.empty() && stk.top().first < temperatures[i]) {
                index = stk.top().second;
                res[index] = (i - index);
                stk.pop();
            }
            stk.push(make_pair(temperatures[i], i));
        }

        return res;
    }
};